Setup initial multi-tenant e-commerce schema, seed data, and workers

PadmanabanMuthukumarasamy opened the pull request:
- Added stores, customers, products, carts, cart_items, orders, and order_items tables
- Implemented multi-tenancy support via CurrentTenant middleware
- Added cart and order controllers with API endpoints
- Implemented CheckoutService for order creation with stock & price validation
- Added Sidekiq workers for inventory reduction and order confirmation emails
- Seeded sample data for adding Customers to Store2

Commit:
Setup initial multi-tenant e-commerce schema, seed data, and workers
- Added stores, customers, products, carts, cart_items, orders, and order_items tables
- Implemented multi-tenancy support via CurrentTenant middleware
- Added cart and order controllers with API endpoints
- Implemented CheckoutService for order creation with stock & price validation
- Added Sidekiq workers for inventory reduction and order confirmation emails
- Seeded sample data for adding Customers to Store2

albert-selvaraj commented on app/controllers/carts_controller.rb:
```diff
@@ -0,0 +1,45 @@
+  before_action :set_cart
+
+  def show
+    render json: serialize_cart(@cart)
```
can we use blueprinter/active model serializer to serialize the hash? we use blueprinter in FS

PadmanabanMuthukumarasamy replied:
I'll add the blueprint in my code and update the **PR**

albert-selvaraj commented on app/Workers/inventory_reduction_worker.rb:
```diff
@@ -0,0 +1,20 @@
+  def self.reduce_stocks_async(order_id)
+    perform_async(order_id)
+  end
+  def perform(order_id)
```
Sidekiq worker can we executed more than once in certain restart behaviour.

albert-selvaraj commented on app/models/product.rb:
```diff
@@ -0,0 +1,9 @@
+  validates :name, :product_code, :price, :stock, presence: true
+  validates :product_code, uniqueness: { scope: :store_id }
+
+  scope :for_current_tenant, -> { where(store_id: CurrentTenant.tenant&.id) }
```
how do we handle current tenant for admin vs customer?

PadmanabanMuthukumarasamy replied:
Currently we’re not using authorization (admin vs customer personas). Our middleware only enforces tenant isolation using X-Tenant header, so for_current_tenant always filters by the tenant.

albert-selvaraj commented on app/services/checkout_service.rb:
```diff
@@ -0,0 +1,43 @@
+    Order.transaction do
+      validate_prices_and_stock!
+
+      order = Order.create!(
```
which tenant the order gets created.

albert-selvaraj commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,13 @@
+
+class OrdersController < ApplicationController
+  def create
+    tenant = CurrentTenant.tenant
```
can we handle this is common middleware, if it is missed, it can create outside the customer account which is critical security issue

PadmanabanMuthukumarasamy replied:
Yes, I’ve updated the logic to always use CurrentTenant.tenant so the order is created within the tenant context. I'll raise one PR for this fix.

Commit:
Ignore tmp folder to clean up PR

albert-selvaraj commented on app/services/checkout_service.rb:
```diff
@@ -0,0 +1,43 @@
+        )
+      end
+
+      @cart.update!(status: "checked_out")
```
use enums for status - rails enum will create helper methods.

albert-selvaraj commented on app/Workers/inventory_reduction_worker.rb:
```diff
@@ -0,0 +1,20 @@
+  include Sidekiq::Worker
+
+  def self.reduce_stocks_async(order_id)
+    perform_async(order_id)
```
how does the account/tenant context set in the worker ? we have to find the order inside tenant scope.

PadmanabanMuthukumarasamy replied:
Right now the workers only get `order_id` so they lose tenant context when **Sidekiq** runs them as we discussed. I’ll fix it by also passing store_id when enqueueing and wrapping the worker logic with `CurrentTenant.tenant = Store.find(store_id).`

albert-selvaraj commented on db/migrate/20250819054807_create_customers.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateCustomers < ActiveRecord::Migration[8.0]
+  def change
+    create_table :customers do |t|
+      t.references :store, null: false, foreign_key: true
```
customers are belong to store right ?

PadmanabanMuthukumarasamy replied:
yes that’s correct.

albert-selvaraj commented on db/cache_schema.rb:
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+ActiveRecord::Schema[7.2].define(version: 1) do
+  create_table "solid_cache_entries", force: :cascade do |t|
```
what is this cache used for ?

PadmanabanMuthukumarasamy replied:
In Rails 8.0 and above, **SolidCable** is the default adapter for ActionCable.

The `solid_cable_messages` table comes from this adapter. It’s used as a cache to persist ActionCable messages in the database instead of Redis so clients can receive broadcasts reliably.

albert-selvaraj reviewed (commented):
Implement Authorization in the product , like Agent and Admin persona

Commit:
integrate Blueprinter and improve tenant handling in workers

- Replaced manual cart serialization with Blueprinter for consistency
- Ensured tenant context is passed in Sidekiq workers (InventoryReductionWorker, OrderConfirmationWorker)
- Updated controllers to consistently use CurrentTenant.tenant for safer multi-tenant handling
